compiler/luci/pass/src/QuantizedModelVerifyPass.cpp

```diff
@@ -0,0 +1,455 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/QuantizedModelVerifyPass.h"
+#include "QuantizationUtils.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/IR/CircleNodeVisitor.h>
+
+// This macro is undef at the end of the file
+#define RETURN_FALSE_UNLESS(ARG) \
+  if (not(ARG))                  \
+  {                              \
+    return false;                \
+  }
+
+namespace
+{
+
+using Granularity = luci::QuantizationGranularity;
+using Type = loco::DataType;
+
+uint32_t rank(const loco::Node *node)
+{
+  auto circle_node = loco::must_cast<const luci::CircleNode *>(node);
+  return circle_node->rank();
+}
+
+bool has_type(const loco::Node *node, Type dtype)
+{
+  auto circle_node = loco::must_cast<const luci::CircleNode *>(node);
+  return circle_node->dtype() == dtype;
+}
+
+bool is_lwq_const(const loco::Node *node)
+{
+  auto circle_node = loco::must_cast<const luci::CircleConst *>(node);
+
+  if (circle_node->quantparam()->scale.size() != 1)
+    return false;
+
+  if (circle_node->quantparam()->zerop.size() != 1)
+    return false;
+
+  return true;
+}
+
+bool is_cwq_const(const loco::Node *node, uint32_t channel_dim)
+{
+  auto circle_node = loco::must_cast<const luci::CircleConst *>(node);
+
+  assert(channel_dim < circle_node->rank()); // FIX_CALLER_UNLESS
+  auto channel_size = circle_node->dim(channel_dim).value();
+
+  if (circle_node->quantparam()->scale.size() != channel_size)
+    return false;
+
+  if (circle_node->quantparam()->zerop.size() != channel_size)
+    return false;
+
+  return true;
+}
+
+struct VerifyQuantizedNodeGranularity final : public luci::CircleNodeVisitor<bool>
+{
+public:
+  VerifyQuantizedNodeGranularity(Granularity gr) : _granularity(gr) {}
+
+private:
+  Granularity _granularity;
+
+private:
+  bool visit(const luci::CircleConv2D *node)
+  {
+    if (_granularity == Granularity::LayerWise)
+    {
+      RETURN_FALSE_UNLESS(is_lwq_const(node->filter()))
+      RETURN_FALSE_UNLESS(is_lwq_const(node->bias()))
+    }
+    else if (_granularity == Granularity::ChannelWise)
+    {
+      RETURN_FALSE_UNLESS(is_cwq_const(node->filter(), 0))
+      RETURN_FALSE_UNLESS(is_cwq_const(node->bias(), rank(node->bias()) - 1))
+    }
+
+    return true;
+  }
+
+  bool visit(const luci::CircleDepthwiseConv2D *node)
+  {
+    if (_granularity == Granularity::LayerWise)
+    {
+      RETURN_FALSE_UNLESS(is_lwq_const(node->filter()))
+      RETURN_FALSE_UNLESS(is_lwq_const(node->bias()))
+    }
+    else if (_granularity == Granularity::ChannelWise)
+    {
+      RETURN_FALSE_UNLESS(is_cwq_const(node->filter(), 3))
+      RETURN_FALSE_UNLESS(is_cwq_const(node->bias(), rank(node->bias()) - 1))
+    }
```

Comment:
(optional) how about spliting `Granularity::LayerWise` and `Granularity::ChannelWise` into to separate visitors?